File: qt/gen_qoffscreensurface.go
package qt

import (
	"unsafe"
)

type QOffscreenSurface struct {
	h          uintptr
	isSubclass bool
}

// NewQOffscreenSurface constructs a new QOffscreenSurface object.
func NewQOffscreenSurface() *QOffscreenSurface {
	g := newQOffscreenSurface(QOffscreenSurface_new())
	g.isSubclass = true
	return g
}

// NewQOffscreenSurface2 constructs a new QOffscreenSurface object.
func NewQOffscreenSurface2(screen *QScreen) *QOffscreenSurface {
	g := newQOffscreenSurface(QOffscreenSurface_new2(screen.cPointer()))
	g.isSubclass = true
	return g
}

// NewQOffscreenSurface3 constructs a new QOffscreenSurface object.
func NewQOffscreenSurface3(screen *QScreen, parent *QObject) *QOffscreenSurface {
	g := newQOffscreenSurface(QOffscreenSurface_new3(screen.cPointer(), parent.cPointer()))
	g.isSubclass = true
	return g
}

func (this *QOffscreenSurface) MetaObject() *QMetaObject {
	return newQMetaObject(QOffscreenSurface_MetaObject(this.h))
}

func (this *QOffscreenSurface) Metacast(param1 string) unsafe.Pointer {
	param1_Cstring := CString(param1)
	defer free(unsafe.Pointer(param1_Cstring))
	return (unsafe.Pointer)(QOffscreenSurface_Metacast(this.h, param1_Cstring))
}

func QOffscreenSurface_Tr(s string) string {
	s_Cstring := CString(s)
	defer free(unsafe.Pointer(s_Cstring))
	var _ms struct_miqt_string = QOffscreenSurface_Tr(s_Cstring)
	_ret := GoStringN(_ms.data, int(int64(_ms.len)))
	free(unsafe.Pointer(_ms.data))
	return _ret
}

func (this *QOffscreenSurface) SurfaceType() SurfaceType {
	xxxxxxxxx
}

func (this *QOffscreenSurface) Create() {
	QOffscreenSurface_Create(this.h)
}

func (this *QOffscreenSurface) Destroy() {
	QOffscreenSurface_Destroy(this.h)
}

func (this *QOffscreenSurface) IsValid() bool {
	return (bool)(QOffscreenSurface_IsValid(this.h))
}

func (this *QOffscreenSurface) SetFormat(format *QSurfaceFormat) {
	QOffscreenSurface_SetFormat(this.h, format.cPointer())
}

func (this *QOffscreenSurface) Format() *QSurfaceFormat {
	_goptr := newQSurfaceFormat(QOffscreenSurface_Format(this.h))
	_goptr.GoGC() // Qt uses pass-by-value semantics for this type. Mimic with finalizer
	return _goptr
}

func (this *QOffscreenSurface) RequestedFormat() *QSurfaceFormat {
	_goptr := newQSurfaceFormat(QOffscreenSurface_RequestedFormat(this.h))
	_goptr.GoGC() // Qt uses pass-by-value semantics for this type. Mimic with finalizer
	return _goptr
}

func (this *QOffscreenSurface) Size() *QSize {
	_goptr := newQSize(QOffscreenSurface_Size(this.h))
	_goptr.GoGC() // Qt uses pass-by-value semantics for this type. Mimic with finalizer
	return _goptr
}

func (this *QOffscreenSurface) Screen() *QScreen {
	return newQScreen(QOffscreenSurface_Screen(this.h))
}

func (this *QOffscreenSurface) SetScreen(screen *QScreen) {
	QOffscreenSurface_SetScreen(this.h, screen.cPointer())
}

func (this *QOffscreenSurface) ScreenChanged(screen *QScreen) {
	QOffscreenSurface_ScreenChanged(this.h, screen.cPointer())
}

func (this *QOffscreenSurface) OnScreenChanged(slot func(screen *QScreen)) {
	QOffscreenSurface_connect_ScreenChanged(this.h, intptr_t(cgo.NewHandle(slot)))
}

//export miqt_exec_callback_QOffscreenSurface_ScreenChanged
func miqt_exec_callback_QOffscreenSurface_ScreenChanged(cb intptr_t, screen *QScreen) {
	gofunc, ok := cgo.Handle(cb).Value().(func(screen *QScreen))
	if !ok {
		panic("miqt: callback of non-callback type (heap corruption?)")
	}

	// Convert all CABI parameters to Go parameters
	slotval1 := newQScreen(screen)

	gofunc(slotval1)
}

func QOffscreenSurface_Tr2(s string, c string) string {
	s_Cstring := CString(s)
	defer free(unsafe.Pointer(s_Cstring))
	c_Cstring := CString(c)
	defer free(unsafe.Pointer(c_Cstring))
	var _ms struct_miqt_string = QOffscreenSurface_Tr2(s_Cstring, c_Cstring)
	_ret := GoStringN(_ms.data, int(int64(_ms.len)))
	free(unsafe.Pointer(_ms.data))
	return _ret
}

func QOffscreenSurface_Tr3(s string, c string, n int) string {
	s_Cstring := CString(s)
	defer free(unsafe.Pointer(s_Cstring))
	c_Cstring := CString(c)
	defer free(unsafe.Pointer(c_Cstring))
	var _ms struct_miqt_string = QOffscreenSurface_Tr3(s_Cstring, c_Cstring, (int)(n))
	_ret := GoStringN(_ms.data, int(int64(_ms.len)))
	free(unsafe.Pointer(_ms.data))
	return _ret
}

func (this *QOffscreenSurface) callVirtualBase_MetaObject() *QMetaObject {
	return newQMetaObject(QOffscreenSurface_virtualbase_MetaObject(unsafe.Pointer(this.h)))
}

func (this *QOffscreenSurface) OnMetaObject(slot func(super func() *QMetaObject) *QMetaObject) {
	if !this.isSubclass {
		panic("miqt: can only override virtual methods for directly constructed types")
	}
	QOffscreenSurface_override_virtual_MetaObject(unsafe.Pointer(this.h), intptr_t(cgo.NewHandle(slot)))
}

//export miqt_exec_callback_QOffscreenSurface_MetaObject
func miqt_exec_callback_QOffscreenSurface_MetaObject(self QOffscreenSurface, cb intptr_t) *QMetaObject {
	gofunc, ok := cgo.Handle(cb).Value().(func(super func() *QMetaObject) *QMetaObject)
	if !ok {
		panic("miqt: callback of non-callback type (heap corruption?)")
	}

	virtualReturn := gofunc((&QOffscreenSurface{h: self}).callVirtualBase_MetaObject)

	return virtualReturn.cPointer()
}

func (this *QOffscreenSurface) callVirtualBase_Metacast(param1 string) unsafe.Pointer {
	param1_Cstring := CString(param1)
	defer free(unsafe.Pointer(param1_Cstring))

	return (unsafe.Pointer)(QOffscreenSurface_virtualbase_Metacast(unsafe.Pointer(this.h), param1_Cstring))
}

func (this *QOffscreenSurface) OnMetacast(slot func(super func(param1 string) unsafe.Pointer, param1 string) unsafe.Pointer) {
	if !this.isSubclass {
		panic("miqt: can only override virtual methods for directly constructed types")
	}
	QOffscreenSurface_override_virtual_Metacast(unsafe.Pointer(this.h), intptr_t(cgo.NewHandle(slot)))
}

//export miqt_exec_callback_QOffscreenSurface_Metacast
func miqt_exec_callback_QOffscreenSurface_Metacast(self QOffscreenSurface, cb intptr_t, param1 *const_char) unsafe.Pointer {
	gofunc, ok := cgo.Handle(cb).Value().(func(super func(param1 string) unsafe.Pointer, param1 string) unsafe.Pointer)
	if !ok {
		panic("miqt: callback of non-callback type (heap corruption?)")
	}
	// Convert all CABI parameters to Go parameters
	param1_ret := param1
	slotval1 := GoString(param1_ret)
	virtualReturn := gofunc((&QOffscreenSurface{h: self}).callVirtualBase_Metacast, slotval1)

	return virtualReturn
}

File: qt/gen_qoffscreensurface.cpp
// +build ignore

#include <QMetaObject>
#include <QObject>
#include <QOffscreenSurface>
#include <QScreen>
#include <QSize>
#include <QString>
#include <QByteArray>
#include <cstring>
#include <QSurface>
#include <QSurfaceFormat>
#include <qoffscreensurface.h>
#include "gen_qoffscreensurface.h"
class MiqtVirtualQOffscreenSurface : public virtual QOffscreenSurface {
public:
MiqtVirtualQOffscreenSurface(): QOffscreenSurface() {};
MiqtVirtualQOffscreenSurface(QScreen* screen): QOffscreenSurface(screen) {};
MiqtVirtualQOffscreenSurface(QScreen* screen, QObject* parent): QOffscreenSurface(screen, parent) {};

virtual ~MiqtVirtualQOffscreenSurface() = default;
// cgo.Handle value for overwritten implementation
	intptr_t handle__MetaObject = 0;
// Subclass to allow providing a Go implementation
	virtual const QMetaObject* metaObject() const override {
if (handle__MetaObject == 0) {
return QOffscreenSurface::metaObject();
}
QMetaObject* callback_return_value = miqt_exec_callback_QOffscreenSurface_MetaObject(const_cast<MiqtVirtualQOffscreenSurface*>(this), handle__MetaObject);

		return callback_return_value;
	}
// Wrapper to allow calling protected method
	QMetaObject* virtualbase_MetaObject() const {

		return (QMetaObject*) QOffscreenSurface::metaObject();

	}
// cgo.Handle value for overwritten implementation
	intptr_t handle__Metacast = 0;
// Subclass to allow providing a Go implementation
	virtual void* qt_metacast(const char* param1) override {
if (handle__Metacast == 0) {
return QOffscreenSurface::qt_metacast(param1);
}
const char* sigval1 = (const char*) param1;

		void* callback_return_value = miqt_exec_callback_QOffscreenSurface_Metacast(this, handle__Metacast, sigval1);

		return callback_return_value;
	}
// Wrapper to allow calling protected method
	void* virtualbase_Metacast(const char* param1) {

		return QOffscreenSurface::qt_metacast(param1);

	}
};
QOffscreenSurface* QOffscreenSurface_new() {
return new MiqtVirtualQOffscreenSurface();
}
QOffscreenSurface* QOffscreenSurface_new2(QScreen* screen) {
return new MiqtVirtualQOffscreenSurface(screen);
}
QOffscreenSurface* QOffscreenSurface_new3(QScreen* screen, QObject* parent) {
return new MiqtVirtualQOffscreenSurface(screen, parent);
}
void QOffscreenSurface_virtbase(QOffscreenSurface* src
, QObject** outptr_QObject
, QSurface** outptr_QSurface
) {
*outptr_QObject = static_cast<QObject*>(src);
*outptr_QSurface = static_cast<QSurface*>(src);
}
QMetaObject* QOffscreenSurface_MetaObject(const QOffscreenSurface* self) {
	return (QMetaObject*) self->metaObject();
}
void* QOffscreenSurface_Metacast(QOffscreenSurface* self, const char* param1) {
	return self->qt_metacast(param1);
}
struct miqt_string QOffscreenSurface_Tr(const char* s) {
	QString _ret = QOffscreenSurface::tr(s);
	// Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
	QByteArray _b = _ret.toUtf8();
	struct miqt_string _ms;
	_ms.len = _b.length();
	_ms.data = static_cast<char*>(malloc(_ms.len));
	memcpy(_ms.data, _b.data(), _ms.len);
	return _ms;
}
SurfaceType QOffscreenSurface_SurfaceType(const QOffscreenSurface* self) {
	return self->surfaceType();
}
void QOffscreenSurface_Create(QOffscreenSurface* self) {
	self->create();
}
void QOffscreenSurface_Destroy(QOffscreenSurface* self) {
	self->destroy();
}
bool QOffscreenSurface_IsValid(const QOffscreenSurface* self) {
	return self->isValid();
}
void QOffscreenSurface_SetFormat(QOffscreenSurface* self, QSurfaceFormat* format) {
	self->setFormat(*format);
}
QSurfaceFormat* QOffscreenSurface_Format(const QOffscreenSurface* self) {
	return new QSurfaceFormat(self->format());
}
QSurfaceFormat* QOffscreenSurface_RequestedFormat(const QOffscreenSurface* self) {
	return new QSurfaceFormat(self->requestedFormat());
}
QSize* QOffscreenSurface_Size(const QOffscreenSurface* self) {
	return new QSize(self->size());
}
QScreen* QOffscreenSurface_Screen(const QOffscreenSurface* self) {
	return self->screen();
}
void QOffscreenSurface_SetScreen(QOffscreenSurface* self, QScreen* screen) {
	self->setScreen(screen);
}
void QOffscreenSurface_ScreenChanged(QOffscreenSurface* self, QScreen* screen) {
	self->screenChanged(screen);
}
void QOffscreenSurface_connect_ScreenChanged(QOffscreenSurface* self, intptr_t slot) {
	MiqtVirtualQOffscreenSurface::connect(self, static_cast<void (QOffscreenSurface::*)(QScreen*)>(&QOffscreenSurface::screenChanged), self, [=](QScreen* screen) {
		QScreen* sigval1 = screen;
		miqt_exec_callback_QOffscreenSurface_ScreenChanged(slot, sigval1);
	});
}
struct miqt_string QOffscreenSurface_Tr2(const char* s, const char* c) {
	QString _ret = QOffscreenSurface::tr(s, c);
	// Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
	QByteArray _b = _ret.toUtf8();
	struct miqt_string _ms;
	_ms.len = _b.length();
	_ms.data = static_cast<char*>(malloc(_ms.len));
	memcpy(_ms.data, _b.data(), _ms.len);
	return _ms;
}
struct miqt_string QOffscreenSurface_Tr3(const char* s, const char* c, int n) {
	QString _ret = QOffscreenSurface::tr(s, c, static_cast<int>(n));
	// Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
	QByteArray _b = _ret.toUtf8();
	struct miqt_string _ms;
	_ms.len = _b.length();
	_ms.data = static_cast<char*>(malloc(_ms.len));
	memcpy(_ms.data, _b.data(), _ms.len);
	return _ms;
}
void QOffscreenSurface_override_virtual_MetaObject(void* self, intptr_t slot) {
	dynamic_cast<MiqtVirtualQOffscreenSurface*>( (QOffscreenSurface*)(self) )->handle__MetaObject = slot;
}
QMetaObject* QOffscreenSurface_virtualbase_MetaObject(const void* self) {
	return ( (const MiqtVirtualQOffscreenSurface*)(self) )->virtualbase_MetaObject();
}
void QOffscreenSurface_override_virtual_Metacast(void* self, intptr_t slot) {
	dynamic_cast<MiqtVirtualQOffscreenSurface*>( (QOffscreenSurface*)(self) )->handle__Metacast = slot;
}
void* QOffscreenSurface_virtualbase_Metacast(void* self, const char* param1) {
	return ( (MiqtVirtualQOffscreenSurface*)(self) )->virtualbase_Metacast(param1);
}
void QOffscreenSurface_Delete(QOffscreenSurface* self, bool isSubclass) {
	if (isSubclass) {
		delete dynamic_cast<MiqtVirtualQOffscreenSurface*>( self );
	} else {
		delete self;
	}
}
